[강진호] Week6

jinhok96 opened the pull request:
## 요구사항

### 기본

- [x] [로그인 페이지] https://bootcamp-api.codeit.kr/docs 에 명세된 “/api/sign-in”으로 { “email”: “test@codeit.com”, “password”: “sprint101” } POST 요청해서 성공 응답을 받고 “/folder”로 이동하나요?

- [x] [회원가입 페이지] 이메일 input에서 “test@codeit.com”으로 “/api/check-email” 이메일 중복 확인 POST 요청하면 이메일 중복 에러를 확인할 수 있나요?

- [x] [기본/회원가입 페이지] 유효한 회원가입 형식의 경우 “/api/sign-up” POST 요청하고 성공 응답을 받으면 “/folder”로 이동하나요?

### 심화

- [x] 로그인/회원가입시 성공 응답으로 받은 accessToken을 로컬 스토리지에 저장했나요?

- [x] 로그인/회원가입 페이지에 접근시 로컬 스토리지에 accessToken이 있는 경우 “/folder” 페이지로 이동하나요?

## 주요 변경사항

- 디코에 작성해주신 예제 참고했습니다!
- 이메일/비밀번호/비밀번호확인 하단 에러 메시지가 focusout 때 출력되는 것과 내부 텍스트는 프로젝트 명세를 반영하였습니다.

## 스크린샷

https://jinhok96-linkbrary.netlify.app/

## 멘토에게

- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.

jinhok96 committed:
refactor: js파일 구조 변경
- utility.js 추가
- signin, signup 페이지 공통 스크립트를 utility.js로 이동
	new file:   js/utility.js
	modified:   signin/index.html
	modified:   signin/index.js

jinhok96 committed:
refactor: js파일 구조 변경
- utility.js 추가
- signin, signup 페이지 공통 스크립트를 utility.js로 이동
	new file:   js/utility.js
	modified:   signin/index.html
	modified:   signin/index.js

jinhok96 committed:
refactor: 파일 이름 변경
chore: 필요없는 <script> 제거
	renamed:    js/utility.js -> js/common_sign.js
	modified:   signin/index.html
	modified:   signin/index.js

jinhok96 committed:
refactor: 파일 이름 변경
chore: 필요없는 <script> 제거
	renamed:    js/utility.js -> js/common_sign.js
	modified:   signin/index.html
	modified:   signin/index.js

jinhok96 committed:
refactor: 파일 이름 변경
chore: 필요없는 <script> 제거
	renamed:    js/utility.js -> js/common_sign.js
	modified:   signin/index.html
	modified:   signin/index.js

jinhok96 committed:
refactor: ACTIVE FUNCTION 위치 재조정
	renamed:    js/common_sign.js -> js/utility.js
	modified:   signin/index.js
	modified:   signup/index.html
	new file:   signup/index.js

jinhok96 committed:
refactor: ACTIVE FUNCTION 위치 재조정
	renamed:    js/common_sign.js -> js/utility.js
	modified:   signin/index.js
	modified:   signup/index.html
	new file:   signup/index.js

jinhok96 committed:
refactor: ACTIVE FUNCTION 위치 재조정
	renamed:    js/common_sign.js -> js/utility.js
	modified:   signin/index.js
	modified:   signup/index.html
	new file:   signup/index.js

jinhok96 committed:
chore: isValidEmail -> emailInput.value 전달 파라미터 추가
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: isValidEmail -> emailInput.value 전달 파라미터 추가
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: 이전 커밋과 동일
	modified:   signup/index.js

jinhok96 committed:
feat: 비밀번호 양식 검사 기능 추가
feat: 이메일 중복 검사 기능 추가
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
feat: 비밀번호 양식 검사 기능 추가
feat: 이메일 중복 검사 기능 추가
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
feat: 비밀번호 양식 검사 기능 추가
feat: 이메일 중복 검사 기능 추가
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
chore: 잘못된 함수 적용 수정
	modified:   signup/index.js

jinhok96 committed:
refactor: UTILITY FUNCTION을 utility.js로 이동
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
refactor: UTILITY FUNCTION을 utility.js로 이동
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
refactor: UTILITY FUNCTION을 utility.js로 이동
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
feat: 비밀번호 확인 유효성 검사 기능 추가
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
feat: 비밀번호 확인 유효성 검사 기능 추가
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
feat: 비밀번호 확인 유효성 검사 기능 추가
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
fix: 잘못된 텍스트 수정
	modified:   signup/index.html

jinhok96 committed:
feat: 회원가입 로직 추가
	modified:   signup/index.js

jinhok96 committed:
chore: eyeBtn 쿼리 수정, eyeBtnCheck 추가
	modified:   js/utility.js
	modified:   signin/index.html
	modified:   signup/index.html
	modified:   signup/index.js

jinhok96 committed:
chore: eyeBtn 쿼리 수정, eyeBtnCheck 추가
	modified:   js/utility.js
	modified:   signin/index.html
	modified:   signup/index.html
	modified:   signup/index.js

jinhok96 committed:
fix: 비밀번호 양식 로직 수정
fix: 로그인, 회원가입 함수 파라미터 수정
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
fix: 비밀번호 양식 로직 수정
fix: 로그인, 회원가입 함수 파라미터 수정
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: isEmpty -> isTextEmpty로 변경
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: isEmpty -> isTextEmpty로 변경
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: isEmpty -> isTextEmpty로 변경
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: 필요없는 함수 제거하고 이름 변경
- isValidEmail 제거
- validateEmail -> isValidEmailFormat으로 변경
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: 필요없는 함수 제거하고 이름 변경
- isValidEmail 제거
- validateEmail -> isValidEmailFormat으로 변경
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: 필요없는 함수 제거하고 이름 변경
- isValidEmail 제거
- validateEmail -> isValidEmailFormat으로 변경
	modified:   js/utility.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: isExistingEmail -> isEmailMatching으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
chore: isExistingEmail -> isEmailMatching으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
chore: isExistingEmail -> isEmailMatching으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
chore: isSamePassword -> isPasswordMatching으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
chore: isSamePassword -> isPasswordMatching으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
chore: isSamePassword -> isPasswordMatching으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
chore: null 케이스를 먼저 처리하도록 순서 변경
	modified:   js/utility.js

jinhok96 committed:
chore: null 케이스를 먼저 처리하도록 순서 변경
	modified:   js/utility.js

jinhok96 committed:
chore: null 케이스를 먼저 처리하도록 순서 변경
	modified:   js/utility.js

jinhok96 committed:
refactor: 필요없는 함수 제거
- isValidPassword 제거
- validatePasswrod -> isValidPasswordFormat으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
refactor: 필요없는 함수 제거
- isValidPassword 제거
- validatePasswrod -> isValidPasswordFormat으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
refactor: 필요없는 함수 제거
- isValidPassword 제거
- validatePasswrod -> isValidPasswordFormat으로 변경
	modified:   js/utility.js
	modified:   signup/index.js

jinhok96 committed:
fix: 누락된 변경사항 수정
	modified:   js/utility.js

jinhok96 committed:
fix: 누락된 변경사항 수정
	modified:   js/utility.js

jinhok96 committed:
fix: 누락된 변경사항 수정
	modified:   js/utility.js

jinhok96 committed:
chore: getUserByLoginInfo 파라미터를 객체로 변경
	modified:   js/utility.js

jinhok96 committed:
chore: getUserByLoginInfo 파라미터를 객체로 변경
	modified:   js/utility.js

jinhok96 committed:
chore: getUserByLoginInfo 파라미터를 객체로 변경
	modified:   js/utility.js

jinhok96 committed:
chore:    js/utility.js -> scripts/utils.js 변경
	renamed:    js/utility.js -> scripts/utils.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore:    js/utility.js -> scripts/utils.js 변경
	renamed:    js/utility.js -> scripts/utils.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore:    js/utility.js -> scripts/utils.js 변경
	renamed:    js/utility.js -> scripts/utils.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: checkLogin에서 fail 케이스를 앞으로 이동
	modified:   signin/index.js

jinhok96 committed:
chore: checkLogin에서 fail 케이스를 앞으로 이동
	modified:   signin/index.js

jinhok96 committed:
chore: 누락된 event.preventDefault 추가
	modified:   signin/index.js

jinhok96 committed:
chore: 누락된 event.preventDefault 추가
	modified:   signin/index.js

jinhok96 committed:
Merge branch 'part1-강진호-week5' of https://github.com/jinhok96/4-Weekly-Mission-jinhok96 into part1-강진호-week5

jinhok96 committed:
Merge branch 'part1-강진호-week6' of https://github.com/jinhok96/4-Weekly-Mission-jinhok96 into part1-강진호-week6

jinhok96 committed:
chore: fail case를 앞으로 순서 이동
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
modified:   scripts/utils.js
	modified:   signin/index.html
	modified:   signin/index.js
	modified:   signup/index.html
	modified:   signup/index.js

jinhok96 committed:
chore: 데이터베이스에서 유저 정보 반환하는 함수 추가
	modified:   scripts/utils.js

jinhok96 committed:
feat: tab 시 btn-eye가 선택되지 않도록 수정
	modified:   signin/index.html
	modified:   signup/index.html

1005hoon commented on signin/index.js:
```diff
@@ -3,28 +3,27 @@ import {
   passwordInput,
   emailErrorMessageElement,
   passwordErrorMessageElement,
```
드는 생각이, dom element selector의 경우, `singin/index.html` 에 있는 돔 요소들만 활용하잖아요. 
이 때, 해당 돔 요소를 담아주는 query selector변수들이 utils.js에 있는거보단, `signin/index.js` 에 있는게 조금 더 말이 될 것 같아요!

아무래도 해당 페이지에서만 존재할 수 있는 돔 요소들이다 보니 공통으로 활용되는 utils.js에 선언시키는건 부적절하게 보입니다.

1005hoon commented on signin/index.js:
```diff
@@ -3,28 +3,27 @@ import {
   hideError,
   isValidUser,
   showPassword,
   hidePassword,
```
조금 더 욕심을 내본다면 ㅋㅋ 
`utils.js` 에는 말 그대로 유틸성 함수들만 두고, 
해당 페이지에만 존재하는 돔 요소를 조작하는 함수들은 `signin/ui-contoller.js` 같은 컨트롤러 파일로 구분해서두거나 

`signin/ui.js` 파일에서 돔 요소 셀렉터들과, `showError` 나 `showPassword` 처럼 해당 돔 요소들을 조작하는 함수들을 모아두고, 
`signin/index.js` 파일에서는 서비스 로직만 구현을 해보는건 어떤가요?

1005hoon commented on signin/index.js:
```diff
@@ -3,28 +3,27 @@ import {
  ********************/
 
 //이메일 에러 검사
 function checkEmailError() {
```
`checkEmailError` 보다는 `validateEmail` 이 조금 더 선언적인 함수 이름일것 같아요!

1005hoon commented on signin/index.js:
```diff
@@ -3,28 +3,27 @@ import {
-  if (!isValidEmail(emailInput.value)) {
+  if (!isValidEmailFormat(emailInput.value)) {
     return showError(emailInput, emailErrorMessageElement, "올바른 이메일 주소가 아닙니다.");
   }
```
조금 더 욕심을 내본다면 emailInput.value를 각 함수마다 전달하는게 아니라 
상단에서 email값을 꺼내놓고 활용하는것도 방법이 될 수 있겠네요! 

```
function validateEmail(e) {
  const email = emailInput.value;

  if (isTextEmpty(email)) { ... }

  if (isValidEmailFormat(email)) { ... }

  return ...
}
```

1005hoon commented on signin/index.js:
```diff
@@ -33,22 +32,28 @@ function checkEmailError() {
   }
 
   return hideError(passwordInput, passwordErrorMessageElement);
 }
```
요녀석도 이메일과 마찬가지로요!

1005hoon commented on signin/index.js:
```diff
@@ -60,23 +65,11 @@ function togglePassword() {
 passwordInput.addEventListener("focusout", checkPasswordError);
-loginBtn.addEventListener("click", checkLogin);
-document.body.addEventListener("keypress", checkLoginByEnter);
+document.querySelector("form").addEventListener("submit", onSubmit);
```
굳! 
다만 노 굳인 부분은 페이지에 폼 요소가 하나만 있다는게 지금은 보장되지만, 프로젝트 규모가 커짐에 따라 해당 쿼리셀렉터는 보장할 수 없습니다.

따라서 emailInput처럼 `const formElement = document.querySelector('.signin-form')` 으로 선택자를 활용해주세요

1005hoon commented on signin/index.js:
```diff
@@ -33,22 +32,28 @@ function checkEmailError() {
+function onSubmit(e) {
+  e.preventDefault();
+  checkSignIn();
 }
```
`onSubmit` 이라는 함수 자체가 사실 폼 제출 시 실행될 비즈니스 로직을 담는다는 함수 이름이기도 하고, 그게 그 함수의 목적이잖아요.

따라서 `checkSignIn` 이라는 함수로 굳이 레이어를 감싸서 호출시킬 필요는 없을것 같아요. 또한 `checkSignIn` 보다는 `loginUser` 라는 함수 이름이 조금 더 적합할것 같구요. 

그래서 조금 더 실무코드적으로 욕심내서 리팩토링 해본다면 이런식으로 구현이 되는게 조금 더 좋을것 같습니다.

```
function handleSubmit(e) {
  e.preventDefault();

  const email = emailInput.value;
  const password = passwordInput.value;

  const user = getUserByEmail(email) // 데이터베이스에 email에 해당되는 사용자가 있다면 사용자 정보를 반환, 없다면 null 반환하는 함수

  if (user === null) {
    // user값이 null일 수 있으니 여기서 널 체크
        throw new ERror("이메일 / 비밀번호 확인 부탁요") 
  }

  // 위에서 null check 했으니, user는 객체임이 분명함! 
  if (user.email !== email) {
    throw new ERror("이메일 / 비밀번호 확인 부탁요") 
  }

  if (user.password !== password) {
    throw new ERror("이메일 / 비밀번호 확인 부탁요") 
  }

  // 우리 데이터베이스에서 반환한 email이 일치하는 사용자 정보와, 사용자가 입력한 이메일 비밀번호가 같으니 로그인 후 절차로 진행
  window.href = ~~~~
}
```

1005hoon commented on signup/index.js:
```diff
@@ -58,11 +57,11 @@ function checkPasswordError() {
 
 //비밀번호 확인 에러 검사
 function checkPasswordCheckError() {
```
validation 함수들의 이름 변경
그리구 입력한 비밀번호들이 일치하는지 체크하는건 사실 단순 string === string 이잖아요. 이걸ㅇ ㅟ해서 함수를 굳이 따로 만드는건 too much일것 같아요!

1005hoon commented on signup/index.js:
```diff
@@ -71,15 +70,20 @@ function checkPasswordCheckError() {
+}
+
+//버튼 클릭 / 인풋 focus 상태에서 엔터 키로 checkSignUp 호출
+function onSubmit(e) {
```
요녀석도 마찬가지일듯 해요! 

```
function handleRegistration(e) {
  e.preventDefault()

  const email = emailInput.value;
  const password = passwordInput.value;
  const checkPassword = checkPasswordInput.value;

  const user = getUserByEmail(email) // 디비에 해당 이메일로 이미 가입된 사용자가 있는지 체크
  if(user) {
    throw new Error("email은 이미 사용중입니다!")
  }

  if (password !== checkPassword) {
    throw new Error("비밀번호 체크 해주세요")
  }

  //  모든 로직 통과했으니  process to signup
  const newUser = createUser({ email, password })
  // 그리고 회원가입 후 절차 진행
}
```

1005hoon commented on signup/index.js:
```diff
@@ -100,25 +104,13 @@ function togglePasswordCheck() {
 passwordCheckInput.addEventListener("focusout", checkPasswordCheckError);
-loginBtn.addEventListener("click", checkSignUp);
-document.body.addEventListener("keypress", checkSignUpByEnter);
+document.querySelector("form").addEventListener("submit", onSubmit);
```
얘도 쿼리셀렉터로!

1005hoon reviewed (changes requested):
엄청 좋습니다! 
이제 정말 실무코드 작성하듯이 로직 구분하는걸 연습해보면 좋겠네요.
디테일한 피드백은 코드단에 남겨두었고, 이 방식으로 활용하면 week6는 이제 로그인 후 절차, 회원가입 후 절차 부분만 수정하면 될 것 같아요.
궁금한점 있다면 디코 콜 주세용

jinhok96 committed:
chore: checkEmailError -> validateEmail로 변경
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: checkPasswordError -> validatePassword로 변경
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: checkPasswordCheckError ->
       validatePasswordCheck로 변경
	modified:   signup/index.js

jinhok96 committed:
chore: signin/ui.js 추가
	new file:   signin/ui.js

jinhok96 committed:
docs: 주석 추가
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
test: ui.js -> ui-sign.js로 변경
- signin, signup에서 공통으로 활용되는 요소 다수 존재
- 따라서 scripts로 이동하고 ui-sign으로 이름 변경
- 해당 페이지에서만 활용되는 스크립트는 각 폴더 안에
  ui.js 또는 index.js에서 활용 예정
	new file:   scripts/ui-sign.js
	deleted:    signin/ui.js

jinhok96 committed:
test: utils-sign.js 추가
- ui 요소를 제외한 로직 관리
	modified:   scripts/ui-sign.js
	new file:   scripts/utils-sign.js
	modified:   scripts/utils.js

jinhok96 committed:
test: 함수 분배
	modified:   scripts/ui-sign.js
	modified:   scripts/utils-sign.js

jinhok96 committed:
test: signup에서만 활용하는 js 추가
	new file:   signup/ui.js
	new file:   signup/utils.js

jinhok96 committed:
refactor: 로직 정리 완료
	modified:   scripts/ui-sign.js
	modified:   scripts/utils-sign.js
	modified:   scripts/utils.js
	modified:   signin/index.js
	modified:   signup/index.js
	modified:   signup/ui.js
	deleted:    signup/utils.js

jinhok96 committed:
chore: email, password 변수에 value 할당
	modified:   signin/index.js

jinhok96 committed:
chore: 이전 커밋 변경사항을 나머지에도 적용
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
refactor: checkSignIn, onSubmit 병합
- handleLogin으로 병합하고 로직 개선
	modified:   signin/index.js

jinhok96 committed:
fix: 잘못된 getUserByEmail 로직 수정
	modified:   signin/index.js

jinhok96 committed:
chore: 필요없는 isPasswordMatching 제거
	modified:   signup/index.js

jinhok96 committed:
refactor: 회원가입 로직을 handleRegister로 재정비
	modified:   scripts/utils-sign.js
	modified:   signin/index.js
	modified:   signup/index.js

jinhok96 committed:
chore: 제거한 함수 로직 수정
	modified:   signup/index.js

jinhok96 committed:
docs: 주석 추가
	modified:   signup/index.js

jinhok96 committed:
chore: 색상 코드 정비
	modified:   css/common.css
	modified:   css/utility.css
	modified:   index.html
	modified:   signin/style.css
	modified:   signup/style.css
	modified:   style.css

jinhok96 committed:
chore: 색상 코드 정리 끝
	modified:   faq/index.html
	modified:   folder/index.html
	modified:   privacy/index.html
	modified:   signin/index.html
	modified:   signup/index.html

jinhok96 commented:
week5 피드백 전부 반영했습니다! week6 내용은 다음 코드리뷰때 올리겠습니다!

1005hoon commented on scripts/utils-sign.js:
```diff
@@ -0,0 +1,32 @@
+}
+
+//이메일이 일치하는 유저 데이터 반환
+function getUserByEmail(emailInputValue) {
```
오늘 마침 디코채널에 올렸는데요! 
좀 더 욕심낸다면 요녀석은 이제 service 로직을 담는 파일로 옮겨봐도 될것 같아요!

1005hoon commented on scripts/utils.js:
```diff
@@ -0,0 +1,20 @@
+ ********************/
+
+//인풋이 비어있는지 검사
+function isTextEmpty(inputValue) {
```
사실 요녀석도 validation의 일환이기는 해서 util에서 빠져도 되지 않을까 생각이 듭니다!

jinhok96 replied:
이 부분은 나중에 검색창이나 폴더명 입력 같은 다른 input 요소를 만들었을 때 활용할 수 있을 것 같아서 utils.js에 배치했습니다!

1005hoon commented on signup/index.js:
```diff
@@ -1,85 +1,154 @@
   }
 
-  if (isExistingEmail(emailInput.value)) {
+  if (email === user.email) {
```
이 로직의 경우, 실제 사용자가 회원가입을 요청 보낸 뒤에 체크하는게 좋을것 같아요. 
실제로도 회원가입 시 클라는 서버에 POST 요청을 보내게 되고, 
서버에서는 데이터베이스를 조회한 후, 만약 요청받은 이메일에 해당하는 사용자가 존재한다면 409 에러를 반환하거든요. 
이 때 받아온 에러를 활용해 이미 사용중인 이메일임을 사용자에게 알려주는 형태로 가보면 어떨까요?

1005hoon commented on signup/index.js:
```diff
@@ -1,85 +1,154 @@
 //비밀번호 확인 에러 검사
-function checkPasswordCheckError() {
-  if (isEmpty(passwordCheckInput.value)) {
+function validatePasswordCheck() {
```
validateCheckPassword가 조금 더 말이 맞을것 가탕요.
그리구 비밀번호 재입력이니까 password-check이라는 이름보다는 confirm password는 어떨까요?

1005hoon commented on signup/index.js:
```diff
@@ -1,85 +1,154 @@
+  const user = getUserByEmail(email);
+
+  //user가 있을 경우 동일한 이메일 존재하므로 회원가입 실패
+  if (user) {
```
여기서 이미 존재하는 이메일인지 여부를 체크하는군요!

1005hoon replied:
그랬을때 모든 validation을 다 떄려 체크하도록 하기보다, 
이 에러케이스에 맞춘 오류 메시지를 던지는것만이 이 함수, if문의 분기의 목적인것 같아요. 이런 형태로 말이죠
```js

  async function handleLogin(e) {
    e.preventDefault();
    const email = emailInput.value;
    const password = passwordInput.value;

    const user = await getUserByEmail(email);

    if (!user) {
      showError(emailInput, emailError, "잘못된 이메일 주소입니다")
      showError(passwordInput, passwordError, "잘못된 비밀번호입니다")
      return
    }

    if (user.password !== password) {
      showError(emailInput, emailError, "잘못된 이메일 주소입니다")
      showError(passwordInput, passwordError, "잘못된 비밀번호입니다")
      return
    }

    window.location.href = "/folders"
  }
```

그리고 작성해주신 validate... 함수들의 경우 말 그대로 data validation차원에서 활용되는 요소들이다 보니 UI validate목저으로 활용되는게 맞을것 같고, 지금처럼 서비스 로직 차원에서의 validation과는 의미가 다른것 같습니다.

1005hoon commented on signup/index.js:
```diff
@@ -1,85 +1,154 @@
+  }
+
+  //비밀번호 형식을 만족하지 않을 경우 회원가입 실패
+  if (!isValidPasswordFormat(password)) {
```
에러 분기를 태운다면 email 값과 password값에 대한 validation들이 먼저 나오고 
그 다음 사용자 조회한 후 진행하는 duplicate과 같은 에러핸들링이 나오는게 조금 더 좋을것 같아요. 
서버와 통신작업이 붙게된다면, client side에 존재하는 값들을 비교하는건 어렵지 않지만 
데이터베이스를 조회해야 하는 작업의 경우 서버통신이라는 비용이 발생하잖아요.
그래서 만약 초반에 서버통신이 발생하지 않아도 캡쳐 할 수 있는 오류들에 대해서는 먼저 캡쳐를 해주는게 좋습니다

1005hoon reviewed (changes requested):
좋슙니다! 
완벽을 향해 달려가고 있군요. 

몇몇가지 에러 분기태우는것과 로직의 관심사의 분리 차원에서 조금 수정하면 좋을듯한 부분들이 있는데, 이 부분들 잘 잡아서 처리해보면 좋을것 같아요. 

디테일한 요소는 코드단의 리뷰로 남겨두었고, 오늘 디코 공유드린 내용 한번 참고해서 업뎃 해보시면 좋을것 같습니다!

jinhok96 committed:
refactor: 파일 이름 변경
	renamed:    signin/index.js -> signin/app.js
	renamed:    scripts/ui-sign.js -> signin/ui.js
	renamed:    signup/index.js -> signup/app.js

jinhok96 committed:
refactor: utils-sign.js 내용을 app.js로 이동
	new file:   scripts/userdb.js
	deleted:    scripts/utils-sign.js
	modified:   signin/app.js
	modified:   signup/app.js

jinhok96 committed:
refactor: 공통으로 사용하는 코드 각 폴더로 복사
	modified:   signin/app.js
	modified:   signup/app.js
	modified:   signup/ui.js

jinhok96 committed:
chore: 잘못된 경로 수정
	modified:   signin/app.js
	modified:   signin/index.html
	modified:   signup/app.js
	modified:   signup/index.html
	modified:   signup/ui.js

jinhok96 committed:
chore: password-check -> confirm-password로 변경
	modified:   signup/app.js
	modified:   signup/index.html
	modified:   signup/ui.js

jinhok96 committed:
chore: 회원가입 로직 재작성
	modified:   signup/app.js

jinhok96 committed:
fix: 회원가입 로직 관련 오류 수정
	modified:   signup/app.js

jinhok96 committed:
chore: fail케이스를 우선 검토하는 로직 적용
	modified:   signup/app.js

jinhok96 committed:
refactor: 비밀번호 토글 함수 ui.js로 이동
refactor: 네트워크 통신이 필요한 로직을 뒤로 이동
	modified:   signin/app.js
	modified:   signin/ui.js
	modified:   signup/app.js
	modified:   signup/ui.js

jinhok96 committed:
chore: 로그인/회원가입 성공 시 링크 이동 전에 에러메시지 제거
	modified:   index.html
	modified:   signin/app.js
	modified:   signup/app.js

jinhok96 committed:
chore: 중복된 로직 제거
	modified:   signup/app.js

jinhok96 committed:
chore: 여러 자잘한 수정사항
refactor: userdb.js -> service.js로 변경
- 공통 서비스 로직 관리
	renamed:    scripts/userdb.js -> scripts/service.js
	modified:   signin/app.js
	modified:   signup/app.js

jinhok96 committed:
feat: api를 이용한 로그인/회원가입 검사 기능 추가
	modified:   scripts/service.js
	modified:   signin/app.js
	modified:   signup/app.js

jinhok96 committed:
chore: 클라이언트 양식 검사 호출 함수 변경
- 특정 함수에서 아이디/비밀번호 검사 통합 관리
	modified:   signin/app.js

jinhok96 committed:
chore: result.error.message 출력 로직 개선
	modified:   signin/app.js
	modified:   signup/app.js

jinhok96 committed:
chore: accessToken 활성화
	modified:   scripts/service.js

jinhok96 committed:
chore: 에러 메시지 코드 개선
	modified:   signin/app.js
	modified:   signup/app.js

1005hoon commented on signin/app.js:
```diff
@@ -0,0 +1,145 @@
+    console.error("에러 발생:", error.message);
+    return null;
+  }
+}
```
@jinhok96 

[제안]
- 여지껏은 우리가 비즈니스 로직과 서버로직을 구분하지 않았었죠 ㅎㅎ 그런데 이제 실제 서버에서 데이터를 가져와야 하는 비지니스 로직이 생겼습니다! 이에 맞추어 코드도 수정되어야 할  것 같아요.
- 지금 이 함수는 실제 로그인 플로우를 다루고 있죠? 따라서 이에 맞춰 함수가 수정이 되어야 할 것 같습니다
```js
async function loginUser(loginInfo) {
  const { email, password } = loginInfo

  const response = await fetch("login url")

  // 여기서 네트워크 오류 발생 여부를 체크합니다.
  // 참고: https://developer.mozilla.org/en-US/docs/Web/API/Response/ok
  if (!response.ok) {
    const errorDetail = await response.text()
    throw new Error(`서버 오류가 발생했습니다: ${errorDetail}`)
  }

  const json = await response.json()
  const { accessToken, user } = json

  return { user, accessToken }
}
```
- 이처럼 사용자 인풋을 받아와 딱 로그인 엔드포인트 요청을 때리고, 그에대한 결과를 반환하는 로직으로 충분할 것 같습니다

1005hoon commented on signin/app.js:
```diff
@@ -0,0 +1,145 @@
+
+  hideError(emailInput, emailErrorMessageElement);
+  return true;
+}
```
@jinhok96 

**[제안]**
- validateEmail의 경우, `emailInput.value` 를 사용하고 있습니다.
- 만약 이메일이 여러 input에 입력이 될 수 있다면 어떨까요? 
- 조금 더 확장성을 고려해두어 `email` 값을 패러미터로 전달받는 형태로 validation 로직만 처리하도록 하구
- 이에 대해 테스트 하는 곳에서 결과값에 따라 show error를 해주는 형태가 조금 더 적절해보입니다

1005hoon commented on signin/app.js:
```diff
@@ -0,0 +1,145 @@
+  hideError(emailInput, emailErrorMessageElement);
+  hideError(passwordInput, passwordErrorMessageElement);
+  return (location.href = "../folder/index.html");
+}
```
@jinhok96 


**[제안]**
- 역시나 이전에는 서비스로직과 비즈니스 로직의 구분이 없이 활용할 수 있었는데, 이제 서버 요청을 실제 하게 되면서 조금 더 분리르 할 수 있게 될것같아요.
- 이 함수의 경우, form 요소에 onSubmit 이벤트가 발생했을 때 실행되는 함수인 만큼 다음과 같이 구분하면 좋겠어요
```js
async function handleSubmit(e) {
  e.preventDefault()

  try {
    // 여기서 로그인 요청 및 에러핸들링 진행
    const data = await loginUser({ email, password })
    
    // 받아온 결과물로 (서비스 로직 호출 결과) 비즈니스 로직 실행 (로컬스토리지 저장 및 이동)
    const { accessToken, user } = data
    localStorage.setItem("token", JSON.stringify(accessToken))
    window.location.href = ""
  } catch (error) {
    // 그래야 실제 발생한 오류를 적절하게 사용자에게 보여줄 수 있을거에요
    console.error(error)
    alert("오류가 발생했어요!")
  }
}


```

1005hoon commented on signup/app.js:
```diff
@@ -0,0 +1,238 @@
+  //1. 클라이언트 회원가입 양식 검사
+  let checkRegisterForm = true;
+
+  if (!(await validateEmail())) {
```
@jinhok96 


**[수정]**
- 최대한 명시적인건 언제나 좋습니다! 
- 해당 함수 호출 결과를 다른 변수에 기록하는게 번거롭더라도, 굳이 변수에 두고 그에대한 if 및 early return 케이스를 활용하는게 유지보수성에도 조금 더 좋을 것 같아요!

1005hoon commented on signup/app.js:
```diff
@@ -0,0 +1,238 @@
+}
+
+//버튼 클릭 / 인풋 focus 상태에서 엔터로 회원가입 로직 실행
+async function handleRegister(e) {
```
@jinhok96 


**[제안]**
- 위에 로그인 플로우와 마찬가지입니다
- 이제 이 녀석은 실제 registration을 처리하기보단, form의 onSubmit 이벤트를 처리하는 목적이겠죠?
- 그리고 실제 회원가입 요청을 서버로 보내는 서비스 로직 함수를 호출하고, 그 결과에 맞춰 사용자에게 ui 업데이트를 해주는 비즈니스 로직을 여기서 활용하면 될 거에요

1005hoon reviewed (approved):
너무 좋습니다! 

항상 빠르게 따라와주셔서 함꼐 진행하는 재미가 있어 즐거웠던 멘토링이였어요. 너무나도 빠르게 학습하시고, 궁금한거 물어보시며 성장하시는게 보여 더 알려드리고 싶은 욕심이 나는 멘토링이였습니다. 

확실히 서버 레이어가 붙기 시작하면서 기존 가볍게 추상화로 구현했던 부분에서 확장하는게 쉽지는 않죠 ㅎㅎ 그래도 관심사의 분리 차원에서 최대한 노력해주신게 보이는 코드에요. 코드 개선할 수 있는 부분들 코드단의 리뷰에 남겨두었습니다.

지난 6주간 너무 고생 많으셨고, 앞으로 남은 코드잇 스프린트 기간동안 더 좋은 성장하셔서 원하는 바 달성하시기를 바랍니다. 고생 많으셨습니다.
